sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt

```diff
@@ -0,0 +1,232 @@
+                .build()
+            )
+          )
+        }
+      }
+    }
+    viewAnnotation.globalLayoutListener = globalLayoutListener
+    inflatedView.viewTreeObserver.addOnGlobalLayoutListener(globalLayoutListener)
+    annotationMap[viewAnnotation.id] = viewAnnotation
+    idLookupMap[inflatedView] = viewAnnotation.id
+    getValue(mapboxMap.addViewAnnotation(viewAnnotation.id, updatedOptions))
+    return inflatedView
+  }
+
+  private fun findByFeatureId(featureId: String): Pair<View?, ViewAnnotationOptions?> {
+    annotationMap.forEach { (id, annotation) ->
+      getValue(mapboxMap.getViewAnnotationOptions(id))?.let { options ->
+        if (options.associatedFeatureId == featureId) {
+          return annotation.view to options
+        }
+      }
+    }
+    return null to null
+  }
+
+  private fun drawAnnotationViews(
+    positionDescriptorCoreList: List<ViewAnnotationPositionDescriptor>
+  ) {
+    // firstly delete views that do not belong to the viewport
+    currentViewsDrawnMap.keys.forEach { id ->
+      if (positionDescriptorCoreList.indexOfFirst { it.identifier == id } == -1) {
+        annotationMap[id]?.let { annotation ->
+          // if view is invisible / gone we don't remove it so that visibility logic could
+          // still be handled by OnGlobalLayoutListener
+          if (annotation.view.visibility == View.VISIBLE) {
+            mapView.removeView(annotation.view)
+          }
+        }
+      }
+    }
+    // add and reposition new and existed views
+    positionDescriptorCoreList.forEach { descriptor ->
+      annotationMap[descriptor.identifier]?.let { annotation ->
+        annotation.viewLayoutParams.apply {
+          width = descriptor.width
+          height = descriptor.height
+        }
+        annotation.view.apply {
+          translationX = descriptor.leftTopCoordinate.x.toFloat()
+          translationY = descriptor.leftTopCoordinate.y.toFloat()
+        }
+        if (!currentViewsDrawnMap.keys.contains(descriptor.identifier) && mapView.indexOfChild(annotation.view) == -1) {
+          mapView.addView(annotation.view, annotation.viewLayoutParams)
+        }
+        hiddenViewMap[annotation.view]?.let { zIndex ->
+          annotation.view.translationZ = zIndex
+          hiddenViewMap.remove(annotation.view)
+        }
+        // as we preserve correct order we bring each view to the front and correct order will be preserved
+        annotation.view.bringToFront()
+      }
+    }
+    // bring to front map view plugins so that they are drawn on top of view annotations
+    viewPlugins.forEach {
+      it.value.bringToFront()
+    }
+    currentViewsDrawnMap.clear()
+    positionDescriptorCoreList.forEach {
+      currentViewsDrawnMap[it.identifier] = it.leftTopCoordinate
+    }
+  }
+
+  private inline fun <reified V> getValue(expected: Expected<String, V>): V? {
+    if (expected.isError) {
+      throw RuntimeException(expected.error)
+    }
+    return expected.value
+  }
```

Comment:
Maybe we need to support this functionality in the base class of Value.
Should we expose an extension function for now? `Value.valueOrThrow()` signature

Reply:
extension functions are sadly not available for Java classes (while `Value` is generated Java class).